src/main/java/picard/util/AlleleSubsettingUtils.java

```diff
@@ -0,0 +1,260 @@
+     * @param snp Snp whose alleles are used for subsetting ctx.
+     * @return a VariantContext subsetted to the alleles in snp, or null if unable.
+     */
+    public static VariantContext subsetVCToMatchSnp(final VariantContext ctx, final Snp snp) {
+        if (ctx.isFiltered()) return null;
+
+        // we can use this VC if it contains both alleles in snp as alleles, one of which is the reference,
+        // or if it contains the NON_REF_ALLELE and one of the snp's alleles, and its reference allele is length 1.
+
+        //TODO: figure out a way to identify the allele corresponding to the snp in the presence of a deletion
+
+        if (ctx.getReference().length() != 1) return null;
+
+        // one of the alleles in snp must match the reference allele
+        final Optional<Byte> referenceAlleleMaybe = Stream.of(snp.getAllele1(), snp.getAllele2())
+                .filter(b -> b == ctx.getReference().getBases()[0]).findAny();
+        if (!referenceAlleleMaybe.isPresent()) return null;
+
+        final byte refAllele = referenceAlleleMaybe.get();
+        final byte otherAllele = snp.getAllele1() == refAllele ? snp.getAllele2() : snp.getAllele1();
+
+        // do we have otherAllele in ctx?
+        final Optional<Allele> altAlleleMaybe = ctx.getAlternateAlleles()
+                .stream()
+                .filter(a -> a.length() == 1 && a.getBases()[0] == otherAllele)
+                .findAny();
+
+        if (altAlleleMaybe.isPresent()) {
+            if (ctx.getAlleles().size() == 2) return ctx;
+
+            return AlleleSubsettingUtils.subsetAlleles(ctx, Arrays.asList(ctx.getReference(), altAlleleMaybe.get()));
+        }
+
+        // if not, perhaps we have NON_REF_ALLELE
+        final Optional<Allele> nonRefAlleleMaybe = ctx.getAlternateAlleles().stream().filter(a -> a.equals(NON_REF_ALLELE)).findAny();
+        if (nonRefAlleleMaybe.isPresent()) {
+
+            final VariantContext vcSubsetted = ctx.getAlleles().size() == 2 ? ctx : AlleleSubsettingUtils.subsetAlleles(ctx, Arrays.asList(ctx.getReference(), nonRefAlleleMaybe.get()));
+            return AlleleSubsettingUtils.swapAlleles(vcSubsetted, NON_REF_ALLELE, Allele.create(otherAllele));
+        }
+        return null;
+    }
+
+    public static VariantContext subsetAlleles(final VariantContext originalVc, final List<Allele> allelesToKeep) {
+        VariantContextBuilder vcBuilder = new VariantContextBuilder(originalVc).alleles(allelesToKeep);
+
+        GenotypesContext newGenotypes = subsetAlleles(originalVc.getGenotypes(), originalVc.getAlleles(), allelesToKeep);
+        vcBuilder.genotypes(newGenotypes);
+        return vcBuilder.make();
+    }
+
+    /**
+     * swaps one of the alleles in a VC (and its genotypes) with another.
+     */
+    public static VariantContext swapAlleles(final VariantContext originalVc, final Allele oldAllele, final Allele newAllele) {
+        if (!originalVc.getAlleles().contains(oldAllele)) return originalVc;
```

Comment:
This seems like it should throw instead of returning.  I'd expect to get back a a variant context that includes the allele I was swapping in.  If you don't think it should change behavior, then expand the doc to mention this case.

Reply:
fixed. and added unit tests.